repoman/flatpak.py

```diff
@@ -0,0 +1,395 @@
+        content_grid.set_margin_top(12)
+        content_grid.set_margin_left(12)
+        content_grid.set_margin_right(12)
+        content_grid.set_margin_bottom(12)
+        content_grid.set_row_spacing(6)
+        content_grid.set_column_spacing(6)
+        content_grid.set_halign(Gtk.Align.CENTER)
+        content_grid.set_hexpand(True)
+        content_area.add(content_grid)
+
+        add_title = Gtk.Label(_("Enter Source Details"))
+        Gtk.StyleContext.add_class(add_title.get_style_context(), "h2")
+        content_grid.attach(add_title, 0, 0, 2, 1)
+
+        self.name_entry = Gtk.Entry()
+        self.name_entry.set_placeholder_text(_("Source Name"))
+        self.name_entry.set_activates_default(True)
+        self.name_entry.connect(_("changed"), self.on_name_entry_changed)
+        self.name_entry.set_width_chars(20)
+        self.name_entry.set_margin_top(12)
+        content_grid.attach(self.name_entry, 0, 1, 1, 1)
+
+        self.url_entry = Gtk.Entry()
+        self.url_entry.set_placeholder_text(_("URL"))
+        self.url_entry.set_activates_default(True)
+        self.url_entry.connect(_("changed"), self.on_url_entry_changed)
+        self.url_entry.set_width_chars(50)
+        self.url_entry.set_margin_top(12)
+        content_grid.attach(self.url_entry, 1, 1, 1, 1)
+
+        self.add_button = self.get_widget_for_response(Gtk.ResponseType.OK)
+        self.add_button.set_sensitive(False)
+
+        self.cancel_button = self.get_widget_for_response(Gtk.ResponseType.CANCEL)
+        self.cancel_button.grab_focus()
+
+        Gtk.StyleContext.add_class(self.add_button.get_style_context(),
+                                   "suggested-action")
+        self.add_button.grab_default()
+
+        self.show_all()
+
+    def on_url_entry_changed(self, entry):
+        entry_text = entry.get_text()
+        entry_valid = flatpak.validate(entry_text)
+        try:
+            self.add_button.set_sensitive(entry_valid)
+        except TypeError:
+            pass
+    
+    def on_name_entry_changed(self, entry, data=None):
+        print(data)
+        # Filter out some special characters, which may not be allowed by spec
+        # and are likely to cause frustration for users who later use the 
+        # Flatpak CLI.
+        replacements = {
+            ' ': '-',
+            '%': '',
+            '(': '',
+            ')': '',
+            '{': '',
+            '}': '',
+            '[': '',
+            ']': '',
+            '#': '',
+            '~': '',
+            '"': '',
+            "'": '',
+            '\\': '',
+            '/': '',
+            '|': '',
+            '&': '',
+            '`': '',
+            '*': '',
+            ';': '',
+            ':': ''
+        }
+        # Store the cursor position so we can return to it later
+        pos = entry.get_position()
+
+        entry_text = entry.get_text()
+        self.log.debug('Validating name: %s, cursor at %s', entry_text, pos)
```

Comment:
Should these debug messages be removed / commented out before merging? I do see them if I run repoman from the terminal.

Reply:
It looks like I was initializing the loggers incorrectly, which was causing the duplication. Fixed in 6702852a44d5dde6de83876130bc8486193e6959